datafusion/sql/src/unparser/expr.rs

```diff
@@ -94,6 +94,7 @@ impl Unparser<'_> {
         Ok(root_expr)
     }
 
+    #[cfg_attr(feature = "recursive_protection", recursive::recursive)]
```

Comment:
This annotation should be enabled by using the  `recursive_protection` feature. I found the CI doesn't enable it. I experimented with it https://github.com/apache/datafusion/pull/15272 (It seems to work well)
You can see how I modified the workflow.